### DISCUSSION TITLE
Possible to set explicit return types on generated hooks with typescript-react-apollo?

### DISCUSSION THREAD

**Discussion by miller-productions** [dotansimha/graphql-code-generator#4052] [General] [answered]
Hi there,

Does anyone know if it is possible to set explicit return types on generated hooks when using [typescript-react-apollo](https://graphql-code-generator.com/docs/plugins/typescript-react-apollo) and setting config `withHooks: true`?

For example, as I currently observe it, a generated mutation hook results in something like...

```
export function useLoginMutation(
  baseOptions?: ApolloReactHooks.MutationHookOptions<
    LoginMutation,
    LoginMutationVariables
  >
) {
  return ApolloReactHooks.useMutation<LoginMutation, LoginMutationVariables>(
    LoginDocument,
    baseOptions
  );
}
```
Note, the return type is missing on the function. If you look into it, it implicitly has a return type of something like `ApolloReactCommon.MutationTuple<LoginMutation, LoginMutationVariables>`.

---
Is there a way of having the generated code explicitly define the return type of the use hook function? E.g. following the same example as above, would look something like...

```
export function useLoginMutation(
  baseOptions?: ApolloReactHooks.MutationHookOptions<
    LoginMutation,
    LoginMutationVariables
  >
): ApolloReactCommon.MutationTuple<LoginMutation, LoginMutationVariables> {
  return ApolloReactHooks.useMutation<LoginMutation, LoginMutationVariables>(
    LoginDocument,
    baseOptions
  );
}
```

**Comment by dotansimha** [comment, marked as answer]
@miller-productions you can use `ReturnType<useLoginMutation>` to get this type, or you can set `withResultType: true` in order to get this generated ;)

**Reply by miller-productions** [reply]
Thanks @dotansimha, much appreciated.

**Comment by Drakota** [comment]
@dotansimha  Not the OP, but I had the same exact issue. I can use the `withResultType: true` option to do something like this, but are there plans to include it automatically without having to do it manually?
![image](https://user-images.githubusercontent.com/22665836/83902068-bf1bf680-a729-11ea-801c-9243f965bbac.png)

**Reply by ardatan** [reply]
I don't think we are planning to add it by default. Why do you need explicit return types? Type should be inferred automatically by TypeScript. Isn't it working as expected?

**Reply by Drakota** [reply]
Currently it doesn't seems to work
![image](https://user-images.githubusercontent.com/22665836/83906641-f8586480-a731-11ea-90ec-1816a08a21b0.png)
I have to manually flag it or else I'm just getting `any` unfortunately

**Reply by dotansimha** [reply]
This seems a bit weird. Do you have `graphql` installed? Which version? (because it requires the typings of it)

We noticed a similar issues a long time ago, in a repositories without `@types/graphql` (it's not needed now, but the missing typings still effect it)

**Reply by miller-productions** [reply]
Hey there, 

Just circling back to this conversation. 

I was also seeing the issue that @Drakota mentioned. i.e. Type wasn't being inferred automatically, was just getting `any`.

However, after updating relevant packages (i.e. graphql, @graphql-codegen, etc) to latest versions typings came through properly and all is working correctly as expected.

✔🆗
